src/provider/store/Store.ts

```diff
@@ -1,122 +1,92 @@
-import {injectable, inject} from 'inversify';
-import {composeWithDevTools} from 'remote-redux-devtools';
-import {Store as ReduxStore, applyMiddleware, createStore, StoreEnhancer, Dispatch, Unsubscribe} from 'redux';
 import {Signal, Aggregators} from 'openfin-service-signal';
 
-import {Inject} from '../common/Injectables';
-import {StoredNotification} from '../model/StoredNotification';
-import {Database, CollectionMap} from '../model/database/Database';
 import {AsyncInit} from '../controller/AsyncInit';
-import {Collection} from '../model/database/Collection';
-
-import {ActionHandlerMap, ActionHandler, RootAction, Action, ActionOf, CustomAction} from './Actions';
-import {RootState} from './State';
-
-export type StoreChangeObserver<T> = (oldValue: T, newValue: T) => void;
 
 /**
  * Subset of properties of `Store` that are safe for use from actions and components.
  */
-export interface StoreAPI {
-    state: RootState;
-    dispatch(action: RootAction): Promise<void>;
+export interface StoreAPI<S, A> {
+    state: S;
+    dispatch(action: A): Promise<void>;
 }
 
-@injectable()
-export class Store extends AsyncInit implements StoreAPI {
-    private static INITIAL_STATE: RootState = {
-        notifications: [],
-        windowVisible: false
-    };
+export abstract class Action<S> {
+    public readonly type?: string;
 
-    public readonly onAction: Signal<[RootAction], Promise<void>> = new Signal(Aggregators.AWAIT_VOID);
+    constructor(type?: string) {
+        this.type = type;
+    }
 
-    private readonly _database: Database;
-    private _actionHandlerMap: ActionHandlerMap;
-    private _store!: ReduxStore<RootState, RootAction>;
+    public reduce(state: S): S {
+        return state;
+    }
+}
 
-    constructor(@inject(Inject.ACTION_HANDLER_MAP) actionHandlerMap: ActionHandlerMap, @inject(Inject.DATABASE) database: Database) {
-        super();
+export abstract class AsyncAction<S> extends Action<S> {
+    public abstract async dispatch(api: StoreAPI<S, Action<S>>): Promise<void>;
+}
+
+export interface Reducer<S> {
+    reduce(state: S): S;
+}
+
+type Listener<S> = (getState: () => S) => void;
+
+export class Store<S, A extends Action<S>> extends AsyncInit implements StoreAPI<S, A> {
```

Comment:
Check class for following public-protected-private ordering

Reply:
Also think we class methods should have return types

Reply:
Apply the same ordering to member variables, so the signal is above everything else. Can also make `_listeners` readonly